server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,321 @@
+                    logger.info("successfully removed decommissioned cluster manager eligible nodes [{}] from voting config, " +
+                        "proceeding to drain the decommissioned nodes", clusterManagerNodesToBeDecommissioned.toString());
+                    registerDecommissionAttribute(decommissionAttribute, listener);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.debug(new ParameterizedMessage(
+                        "failure in removing decommissioned cluster manager eligible nodes from voting config"), exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public AddVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new AddVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    private void clearVotingConfigAfterSuccessfulDecommission() {
+        final ClearVotingConfigExclusionsRequest clearVotingConfigExclusionsRequest = new ClearVotingConfigExclusionsRequest();
+        clearVotingConfigExclusionsRequest.setWaitForRemoval(true);
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            ClearVotingConfigExclusionsAction.NAME,
+            clearVotingConfigExclusionsRequest,
+            new TransportResponseHandler<ClearVotingConfigExclusionsResponse>() {
+                @Override
+                public void handleResponse(ClearVotingConfigExclusionsResponse response) {
+                    logger.info("successfully cleared voting config after decommissioning");
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    logger.debug(new ParameterizedMessage(
+                        "failure in clearing voting config exclusion after decommissioning"), exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public ClearVotingConfigExclusionsResponse read(StreamInput in) throws IOException {
+                    return new ClearVotingConfigExclusionsResponse(in);
+                }
+            }
+        );
+    }
+
+    /**
+     * Registers new decommissioned attribute metadata in the cluster state
+     * <p>
+     * This method can be only called on the cluster-manager node. It tries to create a new decommissioned attribute on the master
+     * and if it was successful it adds new decommissioned attribute to cluster metadata.
+     * <p>
+     * This method should only be called once the eligible cluster manager node having decommissioned attribute is abdicated
```

Comment:
if the existing master was in az to be decommissioned then this code would get triggered on the abdicated master, how are we making sure that this method is being called on the cluster-manager node only? 

Consider the case where the current active master belongs to the to-be-decommissioned zone. Now post the voting exclusion, the current node would stop be the acting active master.

Reply:
That's true. In case, the active leader is set to be decommissioned, this is what would be performed - 
1. Request received on `to-be-decommissioned` active leader
2. Request is validated and then all eligible leaders will be excluded from voting config. This will ensure that the current `to-be-decommissioned` active leader is abdicated. But since the operation has to be performed on a leader node, the service will throw `NotClusterManagerOperation`. An explicit check is also added before we initiate `DECOMMISSION_INIT` and we throw `NotClusterManagerOperation` if the node is not eligible to execute the request.
3. After, the transport action is retried until the `clusterManagerTimeout` is breached. In every retry above checks would be performed.
4. This guarantees that any action after exclusion will always be performed on a active and eligible leader.